### PR TITLE
[misc] minimize main stage

### PR DESCRIPTION
**Pull request by das7pad**
We can reduce the image size of the main stage by 337MB in removing not needed artifacts.

Additionally I am proposing a single COPY statement for the copying of files that are required to build overleaf and another COPY statement for the copying of runtime configuration.
Adding a new file into the docker image is just a matter of creating it in the git repository then.

Changing a settings value in this repository does not invalidate the build cache of overleaf now. That is we can change a setting (a diff of a couple Bytes) and still distribute the same overleaf artifacts (they weight about 1.5GB).

These types of artifacts are not required at run time:
- build dependencies
- git history
- node-gyp build cache
- npm download cache
- v8 cache
- left over temporary directories

With a small hack we can remove the `overleaf/overleaf` git directory as well, we would need to move the `services.js` file to a temporary location, clone the repository and move it back. This has a net saving of 1MB, so I skipped it for now.

---
[1] The image size is measured with `docker save IMAGE_ID | wc --bytes`. I am using the base image and final image with tag `2.0-beta2`. The image `sharelatex/sharelatex:2.0-beta-2` (`sha256:11ac166efa7067e8853fe6b6caa88637035a58a72d1e34c2130f178d85856cf4`) totals `4965017088` Bytes and an image build today using the proposed branch totals `4610853888` Bytes. That's a difference of 337MB.

### CODE REVIEW

**Comment by mserranom** [comment]
> Additionally I am proposing a single COPY statement for the copying of files that are required to build overleaf and another COPY statement for the copying of runtime configuration.

Can you elaborate on this change?

File: Dockerfile
Comment (mserranom) [inline Dockerfile:43]: checkout/install/compile are explicitly split for convenience. These are time consuming steps, and most of the maintenance work involves those stages, hen layer caching helps quite a lot. Unless it saves a significant amount of space they should remain separated.

**Review by mserranom — CHANGES_REQUESTED** [review]
- File renaming should likely be a separate PR
- There are some conflicts to be resolved.

Also left some comments.

File: Dockerfile
Comment (das7pad) [reply Dockerfile:43]:
```
$ sudo docker run --rm --entrypoint sh sharelatex/sharelatex:2.1.1 -c 'find /var/www -name .git | xargs du -h -d 0' | sort -h
216K	/var/www/sharelatex/contacts/.git
260K	/var/www/sharelatex/notifications/.git
320K	/var/www/sharelatex/tags/.git
456K	/var/www/sharelatex/docstore/.git
536K	/var/www/sharelatex/real-time/.git
868K	/var/www/sharelatex/filestore/.git
900K	/var/www/sharelatex/track-changes/.git
964K	/var/www/sharelatex/chat/.git
996K	/var/www/sharelatex/.git
1.1M	/var/www/sharelatex/spelling/.git
1.2M	/var/www/sharelatex/document-updater/.git
2.7M	/var/www/sharelatex/clsi/.git
65M	/var/www/sharelatex/web/.git
```
Does 65MB qualify as 'significant amount of space'?

Comment (mserranom) [reply Dockerfile:43]: Not really enough to sacrifice it in my opinion. However we can perhaps clean up git history just after checking out the repos?

**Comment by das7pad** [comment]
> > Additionally I am proposing a single COPY statement for the copying of files that are required to build overleaf and another COPY statement for the copying of runtime configuration.
> 
> Can you elaborate on this change?

Some config files are required for the build stage, others are required at runtime only.

In splitting the files of this repo into these two categories, we can make better use of layer caching.
E.g. The settings defaults are not required at build time. Upon changing a default, we could rebuild the image, but with this change the layers which build the services can remain the same.

_Also this change reduces the layer count which increases the performance at runtime (lookups in the overlay fs need to check fewer directories)._

File: Dockerfile
Comment (das7pad) [reply Dockerfile:43]: Yes.

Side note: We need to tweak the `git-revision` script in server-pro to amend the output file.

**Comment by das7pad** [comment]
Upon reviewing the change sets, I noticed that we can drop all the npm packages as they are not used any more. These include `grunt` packages, that are not needed since the removal of the `Gruntfile` https://github.com/overleaf/docker-image/commit/27dd97ecc59954a5ad2d8b140f7d5a83073ca2c8 and `simple-git` which is not needed since #133 / https://github.com/overleaf/docker-image/commit/df2d46df829872ccd6d9018dac3f2e6d9b9ae8d0.

**Comment by mserranom** [comment]
I'm happy to merge once:
1. install and compile stages run as separate step (https://github.com/overleaf/docker-image/pull/122/files#r371640805)
2. File renaming is undone (there are actual file reorganisations planned after the monorepo merge)

Once that's done I'll build and test and image prior to merging

**Comment by das7pad** [comment]
Applied review feedback and squashed the commits.

**Review by mserranom — APPROVED** [review]
tested the image, looks all good!

**Pull request merged**
